### ISSUE TITLE
BUG: casting from float32 to uint8 causes different results between s390x and x86_64

### ISSUE THREAD

**Issue by kiszk** [numpy/numpy#23481]
### Describe the issue:

The casting from float32 to uint8 causes ~~incorrect~~ different results for negative (relatively large abs) values (e.g. `-2` and `-1.5`) on s390x. The ~~correct~~ expected behavior is the result generated on x86_64.

This issue is revealed by https://github.com/pytorch/pytorch/issues/97794.

### Reproduce the code example:

```python
import numpy as np
vals = (0, -2, -1.5, -0.5, 0, 0.5, 1.5, 2, 0)
a = np.array(vals, dtype=np.float32).astype(np.uint8)
print(a)
```

### Error message:

```shell
No error message
```

### Runtime information:

s390x
```
% python -c "import sys, numpy; print(numpy.__version__); print(sys.version); print(numpy.show_runtime())"
1.24.2
3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0]
[{'simd_extensions': {'baseline': [],
                      'found': ['VX', 'VXE'],
                      'not_found': ['VXE2']}}]
None
```

x86_64
```
% python -c "import sys, numpy; print(numpy.__version__); print(sys.version); print(numpy.show_runtime())"
1.24.2
3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0]
[{'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2'],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_KNL',
                                    'AVX512_KNM',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}},
 {'architecture': 'Haswell',
  'filepath': '/home/ishizaki/PyTorch/.pytorch-master-build/lib/python3.10/site-packages/numpy.libs/libopenblas64_p-r0-15028c96.3.21.so',
  'internal_api': 'openblas',
  'num_threads': 32,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.21'}]
None
```

### Context for the issue:

When I exec the above program, I got different results on two platforms x86_64 and s390x. I believe that s390x platform generates incorrect result.

on x86_64
```
[  0 254 255   0   0   0   1   2   0]
```

on s390x
```
<string>:1: RuntimeWarning: invalid value encountered in cast
[0 0 0 0 0 0 1 2 0]
```

**Comment by kiszk** [comment]
cc @Andreas-Krebbel

**Comment by mattip** [comment]
What does `sys.byteorder` show on that system?

**Comment by seberg** [comment]
These casts are undefined in C so unfortunately, NumPy cannot (currently?) do more.  We would have to explicitly check the result probably.

On one platform, you get an invalid value warning (the result is out of range for uint8).  On the other, you get the equivalent of something like `arr.astype(np.int64).astype(np.uint8)` for out-of-bound values, and that cast is (at least relatively) well defined, so you get the "correct" wraparound behavior.

Might be interesting if torch does extra work here to ensure the behavior, but compilers can optimize it they way they want, so I would rather assume that is happening.

**Comment by kiszk** [comment]
on x86_64
```
$ python -c "import sys; print(sys.byteorder)"
little
```

on s390x
```
$ python -c "import sys; print(sys.byteorder)"
big
```

**Comment by Andreas-Krebbel** [comment]
As described by @seberg NumPy inherits the behavior from the C compiler and the result is undefined if the number cannot be represented in the target type. On IBM zSystems you can actually get both variants depending on the CPU level you are compiling for.

1. Compiling NumPy with -march=z10 the compiler will emit a cast to signed int first and then to unsigned char. Both steps are well defined in the C standard and do not trigger an exception for the negative inputs from the testcase. However, the result is rather useless. It matches the x86 result though.

2. With IBM z196 we've got an instruction which directly converts to unsigned int. Hence compiling NumPy with -march=z196 (or later) the result is 0 for negative inputs and an exception is raised.

Since the original type cast triggers undefined behavior anyway both variants are ok. We could force 2. by explicitly doing the 2 step cast in NumPy's C code. But this would only be to make the pytorch testcase happy. It might make things slower on some archs. Also I think the second variant is actually slightly more useful since it doesn't make sense to continue computation with the value from 1.

So long story short I agree that we should rather have a look at pytorch here. Either disable the testcase (as it has been done for ARM and Power already), drop the negative values from the list of inputs, or drop the unsigned type from the list of target types.

While I could quickly find the code parts doing the conversion in NumPy I'm still struggling with finding it in PyTorch. What I can say so far is that the behavior in PyTorch does not appear to depend on the -march= setting. So something must be different here.

**Comment by kiszk** [comment]
Thank you for the clarification. I may ask a question about the specification of pytorch conversion in the original issue.

**Comment by seberg** [comment]
As said, this is expected.  We can have an issue about changing it (we probably do, but can expand). This isn't that, so closing.
